app/models/node.rb

```diff
@@ -460,21 +466,25 @@ def location_tags
   # access a tagname /or/ tagname ending in wildcard such as "tagnam*"
   # also searches for other tags whose parent field matches given tagname,
   # but not tags matching given tag's parent field
-  def has_tag(tagname)
-    tags = get_matching_tags_without_aliasing(tagname)
-    # search for tags with parent matching this
-    tags += Tag.includes(:node_tag)
-               .references(:community_tags)
-               .where('community_tags.nid = ? AND parent LIKE ?', id, tagname)
-    # search for parent tag of this, if exists
-    # tag = Tag.where(name: tagname).try(:first)
-    # if tag && tag.parent
-    #  tags += Tag.includes(:node_tag)
-    #             .references(:community_tags)
-    #             .where("community_tags.nid = ? AND name LIKE ?", self.id, tag.parent)
-    # end
-    tids = tags.collect(&:tid).uniq
-    !NodeTag.where('nid IN (?) AND tid IN (?)', id, tids).empty?
+  def has_tag(tagname, preview_tags = [])
+    if preview_tags.blank?
+      tags = get_matching_tags_without_aliasing(tagname)
+      # search for tags with parent matching this
+      tags += Tag.includes(:node_tag)
+                 .references(:community_tags)
+                 .where('community_tags.nid = ? AND parent LIKE ?', id, tagname)
+      # search for parent tag of this, if exists
+      # tag = Tag.where(name: tagname).try(:first)
+      # if tag && tag.parent
+      #  tags += Tag.includes(:node_tag)
+      #             .references(:community_tags)
+      #             .where("community_tags.nid = ? AND name LIKE ?", self.id, tag.parent)
+      # end
+      tids = tags.collect(&:tid).uniq
+      !NodeTag.where('nid IN (?) AND tid IN (?)', id, tids).empty?
+    else
+      match_preview_tag(tagname, preview_tags)
```

Comment:
Here, yes, i see the full "if" loop a bit more clearly than in `has_power_tag` above, appreciate that. But again i wonder if we are overloading it with too much complex functionality. Maybe we could do this at a... "shallower" level and not complicate this very standardized method? I'll keep reading though, thanks!